### ISSUE TITLE
job-manager: posting events from jobtap plugins is unsafe

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#3756]
Recently (#3740) a function was added to allow jobtap plugins to post events for jobs (`flux_jobtap_event_post_pack ()`). However, it turns out that because `event_job_post_pack ()` is called recursively, this is not safe.

Take for example, the following scenario where a jobtap plugin emits a `free` event in the `CLEANUP` state:
```
event_job_post_pack ("finish")
  event_job_update (job)                     // old_state=RUN state=CLEANUP
  jobtap_call ("job.state.cleanup")
  | event_job_post_pack ("free")
  |   event_job_update (job)                 // has_resources = 0
  |   event_job_action (job)         
  |   | event_job_post_pack ("clean")
  |   |   event_job_update (job)             // old_state=CLEANUP state=INACTIVE
  |   |   jobtap_call ("job.state.inactive")
  |   |   event_job_action (job)             // job_decref via zhashx_delete()
  |   |   return
  |   return
  event_job_action (job)                     // deref destroyed job
  ...      
```

Note that because `jobtap_call` occurs before `event_job_action()` not only is `event_job_action()` called on the `INACTIVE` state _before_ the `CLEANUP` state, but the job is destroyed before returning from `jobtap_call()` and the job manager will dereference invalid memory.

Even though `event_job_action()` already calls recursively into `event_job_post_pack()`, I believe this is safe, since `event_job_action()` is the last call in `event_job_post_pack()` (tail recursion). However, jobtap plugins posting events in the middle of the function is not safe.

One workaround for now would be to have the jobtap interface for posting events enqueue the events onto a job-local list, then `event_job_action()` could emit these pending events after it has performed its "action", thereby preserving tail recursion.

**Comment by grondo** [comment]
I took a stab at temporarily queueing events emitted in `event_job_post_pack()` from before the jobtap calls through `event_job_action()` and emitting them at the end of `event_job_post_pack()`, to hopefully preserve the order of events and job state transitions. However, this did not go well, because there are several places in the job manager code that assume an event posted takes immediate effect, e.g.

 * when an `urgency` event is posted, `reprioritize_one()` is called which posts a `priority` event then reorders the alloc queue. If this `priority` event is queued then `job->priority` is updated _asynchronously_ and the synchronous re-order of the alloc queue in the function has no effect
   - we could delay sorting of the alloc queue until the `priority` event is actually processed, but this breaks the case where a plugin wants to reprioritize _all_ jobs. Ideally in this case we could update all job priorities and then sort _once_ after these updates
 * The job manager issues a `depend` anytime `event_job_action()` is called in the DEPEND state. This could result in multiple `depend` events being posted if events are being queued. This could be fixed with a job flag that ensures only one `depend` event is posted at a time.
 * `dependency-add` events cannot be emitted in the NEW state, but once the `submit` event is posted, the job transitions to DEPEND state, and if `dependency-add` events are queued then the `depend` event will be immediately emitted since the unsatisfied dependency count will be 0. I think @garlick suggested we may need a new state here to give plugins a chance to queue and emit dependencies before the DEPEND state, instead just before it. (This worked before because `jobtap_call()` for a state was called before `event_job_action()` for the same state)

**Comment by grondo** [comment]
I believe this was fixed by @garlick in or around de9d43460f5e50bcaef3338b295bb6df0083ccc4
